edb/pgsql/compiler/relgen.py

```diff
@@ -2003,6 +2012,151 @@ def process_set_as_existence_assertion(
+                pgast.ResTarget(
+                    name=alias,
+                    val=pgast.FuncCall(
+                        name=('row_number',),
+                        args=[],
+                        over=pgast.WindowDef(),
+                    )
+                )
+            )
+
+        do_raise = pgast.FuncCall(
+            name=('edgedb', 'raise'),
+            args=[
+                pgast.TypeCast(
+                    arg=pgast.NullConstant(),
+                    type_name=pgast.TypeName(
+                        name=pg_types.pg_type_from_ir_typeref(
+                            ir_arg_set.typeref),
+                    ),
+                ),
+                pgast.StringConstant(val='cardinality_violation'),
+                pgast.NamedFuncArg(
+                    name='msg',
+                    val=pgast.StringConstant(
+                        val='assert_distinct violation: expression returned '
+                            'a set with duplicate elements',
+                    ),
+                ),
+                pgast.NamedFuncArg(
+                    name='constraint',
+                    val=pgast.StringConstant(val='std::assert_distinct'),
+                ),
+            ],
+        )
+
+        check_expr = pgast.CaseExpr(
+            args=[
+                pgast.CaseWhen(
+                    expr=astutils.new_binop(
+                        lexpr=arg_val,
+                        op='IS DISTINCT FROM',
+                        rexpr=pgast.FuncCall(
+                            name=('lag',),
+                            args=[arg_val],
+                            over=pgast.WindowDef(
+                                order_clause=[pgast.SortBy(node=arg_val)],
+                            ),
+                        ),
+                    ),
+                    result=arg_val,
+                ),
+            ],
+            defresult=do_raise,
+        )
+
+        alias2 = ctx.env.aliases.get('v')
+        newctx.rel.target_list.append(
+            pgast.ResTarget(
+                val=check_expr,
+                name=alias2,
+            )
+        )
+
+        pathctx.put_path_var(
+            newctx.rel,
+            ir_set.path_id,
+            check_expr,
+            aspect='value',
+            env=ctx.env,
+        )
+
+        if newctx.rel.sort_clause is None:
+            newctx.rel.sort_clause = []
+        newctx.rel.sort_clause.extend([
+            pgast.SortBy(
+                node=pgast.ColumnRef(name=[sub_rvar.alias.aliasname, alias]),
+            ),
+            pgast.SortBy(
+                node=pgast.ColumnRef(name=[alias2]),
+            ),
+        ])
+
+        pathctx.put_path_id_map(newctx.rel, ir_set.path_id, ir_arg_set.path_id)
+
+    aspects = ('value', 'source')
+
+    func_rvar = relctx.new_rel_rvar(ir_set, newctx.rel, ctx=ctx)
+    relctx.include_rvar(stmt, func_rvar, ir_set.path_id,
+                        pull_namespace=False, aspects=aspects, ctx=ctx)
```

Comment:
Dropping the `pull_namespace=False` fixes the 4 ISE xfails

Reply:
Oh, good.